### PR TITLE
Issue a warning when using zero_grad in DataParallel

### LINKED ISSUE #31768: Exploding gradients when using both backward() and DataParallel
## 🐛 Bug

I am building a model that needs the gradients computed during a backward pass, in the forward of my network. To speed up experimentation I am using nn.DataParallel to spread the workload across multiple GPUs. However, I’m running into a problem of exploding gradients (and losses) when I compute the backward() from the forward() function when I use multiple GPUs.

The following code is a copy of the CIFAR-10 tutorial, except that the forward() function now also does a backward to collect gradients (I would usually do this using backward hooks, but have left out that piece of code for clarity), I have also wrapped the network inside nn.DataParallel

## To Reproduce

To reproduce the behavior:

```import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=4,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)

        self.zero_grad()
        predictions = F.softmax(x).argmax(dim=1)
        ohe = torch.zeros_like(x)
        for i, label in enumerate(predictions):
            ohe[i, label] = 1

        ohe = torch.autograd.Variable(ohe)
        x.backward(gradient=ohe, retain_graph=True)
        self.zero_grad()

        return x


net = Net()
net.cuda()
if torch.cuda.device_count() > 1:
    net = nn.DataParallel(net)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

for epoch in range(2):  # loop over the dataset multiple times

    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data
        inputs, labels = inputs.cuda(), labels.cuda()

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        if i % 2000 == 1999:    # print every 2000 mini-batches
            print('[%d, %5d] loss: %.3f' %
                  (epoch + 1, i + 1, running_loss / 2000))
            running_loss = 0.0

print('Finished Training')
```

## Expected behavior

When I run this code on a single GPU, it works fine (i.e. losses stay in normal range), when I run it on multiple GPUs, the losses explode.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.5 LTS
GCC version: (GCC) 9.2.0
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti
GPU 4: GeForce GTX 1080 Ti
GPU 5: GeForce GTX 1080 Ti

Nvidia driver version: 418.56
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #31768

`DataParallel` creates replicas of the original `nn.Module` with the parameters duplicated onto the destination devices. Calling `backwards` will propagate gradients onto the original module parameters but calling `zero_grad` on the replica module doesn't clear the gradients from the parent module, 

~breaking any model that uses `backward`-`zero_grad` in its `forward`. I fix this by patching the replica module so that `zero_grad` clears grads on the parent as well.~

However, any replica using backwards was broken anyway since the replica's parameters are not leaf nodes in autograd. So, we should raise a warning.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 74286c35:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=74286c35633031dc012e90cf1d6e50167df04529).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4389768) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93795576) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93795576))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/docker/ubuntu/Dockerfile 
Auto-merging .circleci/docker/ubuntu/Dockerfile 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/dimensions.py 
Auto-merging .circleci/cimodel/data/dimensions.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1164,4 +1165,17 @@ def _replicate_for_data_parallel(self):
...
+        old_zero_grad = replica.__class__.zero_grad
+        weak_self = weakref.ref(replica)
+
+        def zero_grad():
+            replica = weak_self()
+            if replica:
+                old_zero_grad(replica)
+            self.zero_grad()
+
+        replica.zero_grad = zero_grad
```
Comment (ezyang) [inline torch/nn/modules/module.py]: I feel like a more normal way to do this is to just bind the function the normal way: `zero_grad.__get__(self, Module)`

**Comment by ezyang** [comment]
Adding albanD to take a look at the test. I didn't realize it from the original bug report, but the use case here is kind of weird.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (peterbell10) [reply torch/nn/modules/module.py]: I've tried your suggestion but get test failures in `test_data_parallel_model_no_refcycles`. Are you sure that bound methods are weak? I can't find any reference to that in the documentation.

**Comment by ErikStammes** [comment]
Thanks @peterbell10 for taking a look at the issue!

@ezyang The use case as I posed it in my bug report is indeed kind of weird, but it is necessary to compute GradCAMs (https://arxiv.org/abs/1610.02391) for input images. It's a relatively common method in explainable AI and weakly supervised detection/localization methods. Maybe that clears things up a little.

**Review by albanD — CHANGES_REQUESTED** [review]
This would be a BC-breaking change right?

Also in DataParallel replica, what happens when you do `.zero_grad()`? The code is a bit obscure as we set in `.replicate()` a Tensor with history in the `_parameters` dictionary. Which break the assumption that parameters are leaf and so `.zero_grad()` won't work. Maybe we want to just through an error saying that `DataParallel` replicas do not hold true Parameters and you should only access the `.grad` fields and `.zero_grad()` using the original network.

**Comment by peterbell10** [comment]
That's a good point: the parameters aren't leaf nodes, so calling backward doesn't really make sense as you can't access the grad anyway. @ErikStammes would you agree with that?

>Also in DataParallel replica, what happens when you do .zero_grad()?

It's essentially a no-op I think. `p.grad` is `None` before and afterwards and no error is raised.

>Maybe we want to just through an error saying that DataParallel replicas do not hold true Parameters and you should only access the .grad fields and .zero_grad() using the original network.

That sounds reasonable although I guess that's still a BC-breaking change. Would a warning be good enough?

**Comment by albanD** [comment]
Warning would be good here I think. Explaining why the `.grad` field won't be updated in `DataParallel`. And advising to the `autograd.grad()` to get gradients.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1164,4 +1166,20 @@ def _replicate_for_data_parallel(self):
...
         replica._buffers = replica._buffers.copy()
         replica._modules = replica._modules.copy()
+
+        # Warn users that gradients don't behave as expected on replica modules
+        old_zero_grad = replica.__class__.zero_grad
+        weak_self = weakref.ref(replica)
+
+        def zero_grad():
+            warnings.warn(
+                "The parameters in data parallel modules are copied from the original module."
```
Comment (albanD) [inline torch/nn/modules/module.py]: nit: Just mention where the warning comes from: "Calling `.zero_grad()` from a module that was passed to a nn.DataParallel() has no effect. The parameters in data parallel modules are copied (in a differentiable manner) from the original module."

**Review by albanD — APPROVED** [review]
Looks good to me.
@ezyang are you satisfied with this version?

File: torch/nn/modules/module.py (hunk 1 above)
Comment (ezyang) [reply torch/nn/modules/module.py]: Well apparently I'm wrong! (I know there must be something weak involved with bound methods, but I guess maybe not with `__get__`

**Review by ezyang — APPROVED** [review]

**Comment by zou3519** [comment]
I am reverting this because it seems to be causing `test_zero_grad (__main__.TestDataParallel)` failures on master (see https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master). Please feel free to resubmit a new PR

File: test/distributed/test_data_parallel.py [hunk 3]
```diff
@@ -630,6 +630,24 @@ def test_replicate_buffers(self):
...
+
+        class Net(torch.nn.Module):
+            def __init__(self, testcase):
+                super(Net, self).__init__()
+                self._testcase = testcase
+
+            def forward(self, x):
+                self._testcase.assertWarnsRegex(
+                    lambda: self.zero_grad(),
+                    "The parameters in data parallel modules are copied from the original module.")
```
Comment (peterbell10) [inline test/distributed/test_data_parallel.py]: The test fails because I changed the error message in the last commit. Fixed in #33064
